diff --git a/nih/watch.c b/nih/watch.c
index d2e4e7f..c930750 100644
--- a/nih/watch.c
+++ b/nih/watch.c
@@ -424,6 +424,13 @@ nih_watch_reader (NihWatch   *watch,
 		if (len < sz)
 			goto finish;
 
+		/* Handle situation where kernel is generating events faster
+		* than can be consumed. Inotify will (somewhat ironically)
+		* generate an event to signify this, so ignore it.
+		*/
+		if (event->mask & IN_Q_OVERFLOW)
+			goto consume;
+
 		/* Find the handle for this watch */
 		handle = nih_watch_handle_by_wd (watch, event->wd);
 		if (handle)
@@ -437,6 +444,7 @@ nih_watch_reader (NihWatch   *watch,
 		if (caught_free)
 			return;
 
+consume:
 		/* Remove the event from the front of the buffer, and
 		 * decrease our own length counter.
 		 */
